src/control/SwerveController.cpp

```diff
@@ -6,54 +6,72 @@
-	if (dtheta != 0) {
-		return {0, {0.0, 0.0, 0.0, 0.0}};
+control::SwerveController::SwerveController(double baseWidth, double baseLength,
+											int epsilon = 1000)
+	: driveMode(DriveMode::Normal), swerve_kinematics(baseWidth, baseLength),
+	  crab_kinematics(baseLength), steer_epsilon(epsilon) {}
+
+swerve_commands_t
+SwerveController::setTurnInPlaceCmdVel(double dtheta, const swerve_rots_t& wheel_rots) const {
+	double dummyScaleFactor;
+	swerve_commands_t commands = setTurnInPlaceCmdVel(dtheta, wheel_rots, dummyScaleFactor);
+	return commands;
+}
+
+swerve_commands_t SwerveController::setTurnInPlaceCmdVel(double dtheta,
+														 const swerve_rots_t& wheel_rots,
+														 double& scaleFactor) const {
+	if (dtheta == 0) {
+		scaleFactor = 0;
+		return {0.0, 0.0, 0.0, 0.0};
 	}
 
-	if (!driveMode.second && !checkWheelRotation(DriveMode::TurnInPlace, wheel_rots))
-		return {0, {0.0, 0.0, 0.0, 0.0}};
+	if (!checkWheelRotation(DriveMode::TurnInPlace, wheel_rots)) {
+		scaleFactor = 0;
+		return {0.0, 0.0, 0.0, 0.0};
+	}
 
 	kinematics::swervewheelvel_t wheelVels =
 		swerve_kinematics.robotVelToWheelVel(0, 0, dtheta);
-	double lfPWM = wheelVels.lfVel / Constants::MAX_WHEEL_VEL;
-	double lbPWM = wheelVels.lbVel / Constants::MAX_WHEEL_VEL;
-	double rfPWM = wheelVels.rfVel / Constants::MAX_WHEEL_VEL;
-	double rbPWM = wheelVels.rbVel / Constants::MAX_WHEEL_VEL;
-	double maxAbsPWM = std::max(std::max(std::abs(lfPWM), std::abs(lbPWM)),
-								std::max(std::abs(rfPWM), std::abs(rbPWM)));
+	double lfPower = wheelVels.lfVel / Constants::MAX_WHEEL_VEL;
+	double lbPower = wheelVels.lbVel / Constants::MAX_WHEEL_VEL;
+	double rfPower = wheelVels.rfVel / Constants::MAX_WHEEL_VEL;
+	double rbPower = wheelVels.rbVel / Constants::MAX_WHEEL_VEL;
+	double maxAbsPWM = std::max(std::max(std::abs(lfPower), std::abs(lbPower)),
+								std::max(std::abs(rfPower), std::abs(rbPower)));
+
 	if (maxAbsPWM > 1) {
-		lfPWM /= maxAbsPWM;
-		lbPWM /= maxAbsPWM;
-		rfPWM /= maxAbsPWM;
-		rbPWM /= maxAbsPWM;
+		lfPower /= maxAbsPWM;
+		lbPower /= maxAbsPWM;
+		rfPower /= maxAbsPWM;
+		rbPower /= maxAbsPWM;
+		scaleFactor = maxAbsPWM;
+	} else {
+		scaleFactor = 1.0;
 	}
 
-	return {maxAbsPWM > 1 ? maxAbsPWM : 1.0, {lfPWM, rfPWM, lbPWM, rbPWM}};
+	return {lfPower, rfPower, lbPower, rbPower};
+}
+
+swerve_commands_t SwerveController::setCrabCmdVel(double dtheta, double dy,
+												  const swerve_rots_t& wheel_rots) const {
+	double dummyScaleFactor;
+	swerve_commands_t commands = setCrabCmdVel(dtheta, dy, wheel_rots, dummyScaleFactor);
+	return commands;
 }
 
-std::pair<double, swerve_commands_t>
-SwerveController::setCrabCmdVel(double dtheta, double dy, swerve_rots_t wheel_rots) {
+swerve_commands_t SwerveController::setCrabCmdVel(double dtheta, double dy,
+												  const swerve_rots_t& wheel_rots,
+												  double& scaleFactor) const {
 	if (Globals::E_STOP && (dtheta != 0 || dy != 0)) {
```

Comment:
No point in checking the `E_STOP` value, since this method isn't responsible for policing that. Also means we can cut the dependency on globals

Reply:
As an aside, we can just cut this whole special case, since if the inputs are zero then the outputs will be zero.